src/client.py

```diff
@@ -1,45 +1,83 @@
 from . import errors, defines
 from .block import Block
 from .connection import Connection
+from .progress import Progress
 from .protocol import ServerPacketTypes
 from .util.escape import escape_params
 from .util.helpers import chunks
 
 
 class QueryResult(object):
-    def __init__(self, with_column_types=False):
+    def __init__(
+            self, packet_generator,
+            with_column_types=False, columnar=False):
+        self.packet_generator = packet_generator
         self.with_column_types = with_column_types
 
         self.data = []
         self.columns_with_types = []
+        self.columnar = columnar
 
         super(QueryResult, self).__init__()
 
+    def store(self, packet):
+        block = getattr(packet, 'block', None)
+        if block is None:
+            return
+
+        # Header block contains no rows. Pick columns from it.
+        if block.rows:
+            if self.columnar:
+                columns = block.get_columns()
+                if self.data:
+                    # Extend corresponding column.
+                    for i, column in enumerate(columns):
+                        self.data[i] += column
+                else:
+                    self.data.extend(columns)
+            else:
+                self.data.extend(block.get_rows())
+
+        elif not self.columns_with_types:
+            self.columns_with_types = block.columns_with_types
+
     def get_result(self):
+        for packet in self.packet_generator:
+            self.store(packet)
+
         if self.with_column_types:
             return self.data, self.columns_with_types
         else:
             return self.data
 
 
-class Progress(object):
-    def __init__(self, client_connection, result, progress_gen):
-        self.connection = client_connection
-        self.result = result
-        self.progress_gen = progress_gen
+class ProgressQueryResult(QueryResult):
+    def __init__(
+            self, packet_generator,
+            with_column_types=False, columnar=False):
+        self.progress_totals = Progress()
 
-        super(Progress, self).__init__()
+        super(ProgressQueryResult, self).__init__(
+            packet_generator, with_column_types, columnar
+        )
+
+    def store_progress(self, progress_packet):
+        self.progress_totals.rows += progress_packet.rows
+        self.progress_totals.bytes += progress_packet.bytes
+        self.progress_totals.total_rows += progress_packet.total_rows
+        return self.progress_totals.rows, self.progress_totals.total_rows
 
     def __iter__(self):
         return self
 
     def next(self):
-        try:
-            return next(self.progress_gen)
-        except Exception as ex:
-            if not isinstance(ex, StopIteration):
-                self.connection.disconnect()
-            raise
+        while True:
+            packet = next(self.packet_generator)
```

Comment:
We should `.disconnect()` if non-`StopIteration` exception raises and re-raise this exception. See previous code. This in the `packet_generator` now. There is no need in disconnect on `StopIteration`

Reply:
Not 100% sure what you mean here. By moving the exception handler with the `disconnect()` call down to `packet_generator`, I avoided the need to add a case for `StopIteration`, as the `StopIteration` would now only be raised in `ProgressQueryResult.next()`, where we do not need to know about the connection or disconnect it. 

I _think_ the overall behavior is the same as it used to be, but I could be missing something.